File: 2021/elixir/day13.exs
#! /usr/bin/env elixir

defmodule Day13 do
  def parse(input) do
    input
    |> Stream.map(&String.trim/1)
    |> Stream.map(&parse_row/1)
    |> Enum.reject(&is_nil/1)
  end

  def parse_row(""), do: nil

  def parse_row(<<"fold along ", cmd::binary>>) do
    [f, n] = String.split(cmd, "=")
    {:cmd, String.to_atom(f), String.to_integer(n)}
  end

  def parse_row(row) do
    [x, y] = row |> String.split(",") |> Enum.map(&String.to_integer/1)
    {:dot, {x, y}}
  end

  def fold({:dot, key}, dots) do
    Map.put(dots, key, :"#")
  end

  def fold({:cmd, :y, n}, dots) do
    out = Map.get(dots, :out, false)

    dots
    |> Map.delete(:out)
    |> Map.keys()
    |> Enum.map(fn {x, y} ->
      (y < n && {x, y}) || {x, 2 * n - y}
    end)
    |> Enum.map(&{&1, :"#"})
    |> Enum.into(%{})
    |> maybe_print_res1(out)
  end

  def fold({:cmd, :x, n}, dots) do
    out = Map.get(dots, :out, false)

    dots
    |> Map.delete(:out)
    |> Map.keys()
    |> Enum.map(fn {x, y} ->
      (x < n && {x, y}) || {2 * n - x, y}
    end)
    |> Enum.map(&{&1, :"#"})
    |> Enum.into(%{})
    |> maybe_print_res1(out)
  end

  def fold(r, _dots) do
    raise "unknown row to fold: #{r.inspect}"
  end

  def maybe_print_res1(dots, true) do
    dots |> Enum.count() |> IO.inspect(label: "res1")
    dots
  end

  def maybe_print_res1(dots, false), do: dots

  def get_size(dots) do
    keys = Map.keys(dots)
    {x, _} = Enum.max_by(keys, fn {x, _y} -> x end)
    {_, y} = Enum.max_by(keys, fn {_x, y} -> y end)

    {x, y}
  end
end

# input =
# """
# 6,10
# 0,14
# 9,10
# 0,3
# 10,4
# 4,11
# 6,0
# 6,12
# 4,1
# 0,13
# 10,12
# 3,4
# 3,0
# 8,4
# 1,10
# 2,14
# 8,10
# 9,0

# fold along y=7
# fold along x=5
# """
# |> String.split("\n")

input = "../inputs/day13" |> File.stream!()

msg =
  input
  |> Day13.parse()
  |> Enum.reduce(%{out: true}, &Day13.fold/2)

{x, y} = Day13.get_size(msg)

IO.puts("res2: ")

# EFJKZLBL
Enum.each(0..y, fn y ->
  Enum.map(0..x, fn x ->
    Map.get(msg, {x, y}, " ")
  end)
  |> Enum.join()
  |> IO.puts()
end)

File: 2021/ruby/day13.rb
#! /usr/bin/env ruby
# require "set"
require_relative "nanotest"

# input = <<INPUT
# 6,10
# 0,14
# 9,10
# 0,3
# 10,4
# 4,11
# 6,0
# 6,12
# 4,1
# 0,13
# 10,12
# 3,4
# 3,0
# 8,4
# 1,10
# 2,14
# 8,10
# 9,0

# fold along y=7
# fold along x=5
# INPUT

input = File.read("../inputs/day13")

VERBOSE = ARGV.delete("-v")

def prep_data(input)
  data, cmds = input.strip.split("\n\n")
  dots = {}

  data.split("\n").each do |r|
    k = r.split(",").map(&:to_i)
    dots[k] = true
  end

  cmds = cmds.split("\n").map do |r|
    f, n = r.split(" ").last.split("=")
    [f.to_sym, n.to_i]
  end

  [dots, cmds]
end

def fold(d, cmd)
  dir, n = cmd

  d.keys.reduce({}) do |acc, k|
    x, y = k

    kk = case dir
      when :x
        x < n ? [x, y] : [2 * n - x, y]
      when :y
        y < n ? [x, y] : [x, 2 * n - y]
      end
    acc[kk] = true
    acc
  end
end

def pmax(dots)
  kk = dots.keys
  xm = kk.map { |x, y| x }.max
  ym = kk.map { |x, y| y }.max
  [xm, ym]
end

if $0 == __FILE__
  Utils.day(2021, 13)
  dots, cmds = prep_data(input)

  # T1
  t1 = fold(dots, cmds[0])
  puts "res1:", "(after #{cmds[0]}) #{t1.size}"

  # T2
  dots = cmds.reduce(dots) { |acc, cmd| fold(acc, cmd) }

  xx, yy = pmax(dots)
  puts "res2:"

  (0..yy + 1).each do |y|
    r = (0..xx + 1).map do |x|
      dots.has_key?([x, y]) ? "#".white : " "
    end.join
    print("#{r}\n")
  end
end
